File: 2022/20221028/keypad.js
function solution(numbers, hand) {
  numbers = numbers.map((number) => {
    if (number === 0) {
      return 11;
    }
    return number;
  });

  const position = [];
  let leftPosition = 10;
  let rightPosition = 12;

  numbers.forEach((number) => {
    if (number % 3 === 1) {
      leftPosition = number;
      position.push('L');
      return;
    }

    if (number % 3 === 0) {
      rightPosition = number;
      position.push('R');
      return;
    }

    const leftDistance = calculateDistance(leftPosition, number);
    const rightDistance = calculateDistance(rightPosition, number);

    if (leftDistance > rightDistance) {
      rightPosition = number;
      return position.push('R');
    }

    if (leftDistance < rightDistance) {
      leftPosition = number;
      return position.push('L');
    }

    if (leftDistance === rightDistance) {
      if (hand === 'right') {
        rightPosition = number;
        return position.push('R');
      }
      if (hand === 'left') {
        leftPosition = number;
        return position.push('L');
      }
    }
  });
  return position.join('');
}

const calculateDistance = (position, targetPosition) => {
  let count = 0;

  while (position !== targetPosition) {
    if (position + 1 === targetPosition || position - 1 === targetPosition) {
      count += 1;
      return count;
    }

    if (position > targetPosition) {
      position -= 3;
      count += 1;
      continue;
    }

    if (position < targetPosition) {
      position += 3;
      count += 1;
      continue;
    }
  }
  return count;
};

File: 2022/20221028/keypad/src/main/java/KeyPad.java
public class KeyPad {
    private int left;
    private int right;

    public String solution(int[] numbers, String hand) {
        String answer = "";
        left = 10;
        right = 12;

        for (int number : numbers) {
            answer += getPressedHand(number, hand);
        }

        return answer;
    }

    public String getPressedHand(int number, String hand) {
        if (number == 0) {
            number = 11;
        }

        if (number % 3 == 1) {
            left = number;
            return "L";
        }

        if (number % 3 == 0) {
            right = number;
            return "R";
        }

        if (number % 3 == 2) {
            int leftGap = Math.abs(left - number) / 3 + Math.abs(left - number) % 3;
            int rightGap = Math.abs(right - number) / 3 + Math.abs(right - number) % 3;

            if (leftGap < rightGap) {
                left = number;
                return "L";
            }

            if (rightGap < leftGap) {
                right = number;
                return "R";
            }

            if (hand.equals("right")) {
                right = number;
                return "R";
            }

            if (hand.equals("left")) {
                left = number;
                return "L";
            }
        }
        return "";
    }
}
